File: frontend/src/constants/languages.ts
export const LANGUAGES = [
  { code: "en", name: "English" },
  { code: "zh", name: "中文" },
  { code: "sv", name: "Svenska" },
  { code: "ja", name: "日本語" },
  { code: "ko", name: "한국어" },
  { code: "fr", name: "Français" },
  { code: "de", name: "Deutsch" },
  { code: "es", name: "Español" },
] as const;

export type LanguageCode = (typeof LANGUAGES)[number]["code"];

File: backend/app/core/languages.py
import re

DEFAULT_LANGUAGE = "en"
SUPPORTED_LANGUAGES = {
    "en": "English",
    "zh": "中文",
    "sv": "Svenska",
    "ja": "日本語",
    "ko": "한국어",
    "fr": "Français",
    "de": "Deutsch",
    "es": "Español",
}


def normalize_language(value: str | None, default: str = DEFAULT_LANGUAGE) -> str:
    language = (value or default).strip().lower().replace("_", "-")
    if not re.fullmatch(r"[a-z]{2,3}(?:-[a-z]{2,4})?", language):
        raise ValueError("language must be a valid language tag such as en, zh, or sv")
    return language
